File: src/DashSpec.Core/Parsing/InspectPresentationParser.cs
using DashSpec.Core.Model;
using DashSpec.Core.Runtime;

namespace DashSpec.Core.Parsing;

internal static class InspectPresentationParser
{
    public static InspectPresentation Parse(TokenReader reader, string context)
    {
        BlockSyntax.BeginBlock(reader);
        reader.SkipNewlines();

        string? tooltipId = null;
        string? label = null;
        string? format = null;
        string? split = null;

        while (!BlockSyntax.IsBlockEnd(reader, "inspect") && !reader.IsEof)
        {
            reader.SkipNewlines();
            if (BlockSyntax.IsBlockEnd(reader, "inspect"))
            {
                break;
            }

            if (reader.TryKeyword("use"))
            {
                if (!reader.TryKeyword("tooltip"))
                {
                    throw new DashSpecParseException($"{context}: inspect 'use' requires 'tooltip <id>'.");
                }

                tooltipId = reader.ReadIdent();
                if (string.IsNullOrWhiteSpace(tooltipId))
                {
                    throw new DashSpecParseException($"{context}: inspect use tooltip requires an id.");
                }

                continue;
            }

            if (reader.TryKeyword("label"))
            {
                reader.Expect(TokenKind.Eq);
                label = reader.ReadString();
                continue;
            }

            if (reader.TryKeyword("as"))
            {
                var token = reader.ReadIdent();
                format = token?.ToLowerInvariant() switch
                {
                    "list" or "bullets" or "ul" => "list",
                    "inline" or "line" or "text" => "inline",
                    _ => throw new DashSpecParseException(
                        $"{context}: inspect as must be list or inline; got '{token}'."),
                };
                continue;
            }

            if (reader.TryKeyword("split"))
            {
                reader.Expect(TokenKind.Eq);
                split = reader.ReadString();
                continue;
            }

            throw reader.Unexpected();
        }

        BlockSyntax.ExpectBlockEnd(reader, "inspect");

        if (string.IsNullOrWhiteSpace(tooltipId))
        {
            throw new DashSpecParseException($"{context}: inspect requires 'use tooltip <id>'.");
        }

        var resolvedFormat = format
            ?? (!string.IsNullOrWhiteSpace(split) ? "list" : "inline");

        return new InspectPresentation(
            tooltipId,
            label,
            resolvedFormat,
            string.IsNullOrWhiteSpace(split) ? ", " : split);
    }

    public static InspectPresentation? Merge(InspectPresentation? left, InspectPresentation? right)
    {
        if (right is null)
        {
            return left;
        }

        if (left is null)
        {
            return right;
        }

        return new InspectPresentation(
            right.TooltipId ?? left.TooltipId,
            right.Label ?? left.Label,
            right.Format ?? left.Format,
            right.Split != ", " ? right.Split : left.Split);
    }

    public static TooltipFormat ToTooltipFormat(InspectPresentation? inspect) =>
        TooltipFormatParser.Parse(inspect?.Format, TooltipFormat.Inline);
}

File: src/DashSpec.Modeling.Parse/Diagram/InspectPresentationParser.fs
namespace DashSpec.Modeling.Parse.Diagram

open System
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Lexing

module InspectPresentationParser =

    let parse (reader: TokenReader) (context: string) =
        BlockSyntax.beginBlock reader
        reader.SkipNewlines()

        let mutable tooltipId: string option = None
        let mutable label: string option = None
        let mutable format: string option = None
        let mutable split: string option = None

        while not (BlockSyntax.isBlockEnd reader "inspect" None) && not reader.IsEof do
            reader.SkipNewlines()
            if BlockSyntax.isBlockEnd reader "inspect" None then ()
            elif reader.TryKeyword "use" then
                if not (reader.TryKeyword "tooltip") then
                    raise (DashSpecParseException($"{context}: inspect 'use' requires 'tooltip <id>'."))
                let id = reader.ReadIdent()
                if String.IsNullOrWhiteSpace id then
                    raise (DashSpecParseException($"{context}: inspect use tooltip requires an id."))
                tooltipId <- Some id
            elif reader.TryKeyword "label" then
                reader.Expect TokenKind.Eq
                label <- Some(reader.ReadString())
            elif reader.TryKeyword "as" then
                let token = reader.ReadIdent()
                format <-
                    match token.ToLowerInvariant() with
                    | "list" | "bullets" | "ul" -> Some "list"
                    | "inline" | "line" | "text" -> Some "inline"
                    | _ ->
                        raise (DashSpecParseException($"{context}: inspect as must be list or inline; got '{token}'."))
            elif reader.TryKeyword "split" then
                reader.Expect TokenKind.Eq
                split <- Some(reader.ReadString())
            else
                raise (reader.Unexpected())

        BlockSyntax.expectBlockEnd reader "inspect" None

        if tooltipId.IsNone || String.IsNullOrWhiteSpace tooltipId.Value then
            raise (DashSpecParseException($"{context}: inspect requires 'use tooltip <id>'."))

        let resolvedFormat =
            match format with
            | Some f -> f
            | None when split.IsSome -> "list"
            | None -> "inline"

        let resolvedSplit =
            match split with
            | Some s when not (String.IsNullOrWhiteSpace s) -> s
            | _ -> ", "

        { TooltipId = tooltipId
          Label = label
          Format = resolvedFormat
          Split = resolvedSplit }

    let merge (left: InspectPresentation option) (right: InspectPresentation option) =
        match right with
        | None -> left
        | Some r ->
            match left with
            | None -> Some r
            | Some l ->
                Some
                    { TooltipId = r.TooltipId |> Option.orElse l.TooltipId
                      Label = r.Label |> Option.orElse l.Label
                      Format = if String.IsNullOrWhiteSpace r.Format then l.Format else r.Format
                      Split = if r.Split <> ", " then r.Split else l.Split }
